cli/spec.py

```diff
@@ -0,0 +1,260 @@
+    :param trim: specifies whether the certain option
+        parameters should be removed
+    """
+    global_options = {}
+    subcommand_options = {}
+    _process_flat_options(global_options, spec_dict, lookup, trim)
+    _process_groups_options(global_options, spec_dict, lookup, trim)
+    # collect sub parser global_options
+    if 'subparsers' in spec_dict:
+        for subparser_name, params in spec_dict['subparsers'].iteritems():
+            subcommand_options[subparser_name], _ = _process_options(
+                params, lookup=True, trim=True)
+
+    return global_options, subcommand_options
+
+
+def _process_groups_options(options, spec_dict, lookup, trim):
+    """
+    Gets the dict of options nested within the spec groups
+
+    :param options:  the current list of options
+    :param spec_dict: the dictionary to look for new group options.
+    :param lookup: specifies whether the options lookup is required.
+    :param trim: specifies whether the certain option
+        parameters should be removed
+    """
+    if 'groups' in spec_dict:
+        for group in spec_dict['groups']:
+            _process_flat_options(options, group, lookup, trim)
+
+
+def _process_flat_options(options, spec_dict, lookup, trim):
+    """
+    Gest the dict of options listed in the spec of group.
+
+    This method will also remove some methods and will replace
+     __<value>__ pattens in the option properties (e.g. __default__, __FILE__)
+
+    :param options:  the current list of options
+    :param spec_dict: the dictionary to look for new options.
+    :param lookup: specifies whether the options lookup is required.
+    :param trim: specifies whether the certain option
+        parameters should be removed
+    """
+    if 'options' in spec_dict:
+        for opt_name, opt_params in spec_dict['options'].iteritems():
+            if lookup:
+                _lookup_option(opt_name, opt_params)
+
+            # get a parameters copy with all the keys.
+            options[opt_name] = dict(opt_params)
+
+            if trim:
+                _trim_option(opt_name, opt_params)
+
+
+def _lookup_option(opt_name, opt_params):
+    """
+    Modifies existing option parameters by replaceing ___*___ patterns
+      and inserting default values into the help.
+    :param opt_name: the option name to process
+    :param opt_params: the dictionary of the option parameters
+        (help, type, default, etc)
+    """
+
+    # first check __*__ pattern
+    for opt_param_name, param_value in opt_params.iteritems():
+        for lookup_string, lookup_key in LOOKUP_DICT.iteritems():
+            if lookup_string in str(param_value):
+                opt_params[opt_param_name] = \
+                    param_value.replace(lookup_string, str(opt_params.get(
+                        lookup_key, lookup_string)))
+
+    # insert default value into help.
+    if 'help' in opt_params and 'default' in opt_params:
```

Comment:
Please remember that we're using the 'clg' module, user's may add '__DEFAULT__' in option's help. If a user already provided the '__DEFAULT__' in help, there is not need to add this addition.